ramips: add support for D-Link DIR-867 A1 and DIR-882 A1

mbc07 opened the pull request:
This PR continues the work started in PR #2588 and #3029 by adding support for DIR-867 A1 and DIR-882 A1. There are several other routers that are based on the same base board (DIR-1760/DIR-1960/DIR-2660 from D-Link and MR2600 from Motorola), but since they require more changes due to different flash layout and minor differences on the LEDs, and considering that AFAICT @Lucky1openwrt is already working on the DIR-1960, I focused this PR on what I can test at the moment, the DIR-867/878/882 trio.

The existing DTS for the DIR-878 was changed to a common DTS and a separate DTS with the differences were made for the three targets (DIR-867/878/882). I carefully tested all of them and the compiled images works fine, the devices are properly identified, all LEDs working as they should and the USB ports works out of the box (on DIR-882 case). DIR-867 is hardware limited to 3x3 streams but this didn't require any specific change in the DTS, as the mt76 driver seems to handle that by itself (by probing the EEPROM apparently? Not entirely sure).

Anyway, this PR is ready for review.

Lucky1openwrt commented:
> and considering that AFAICT @Lucky1openwrt is already working on the DIR-1960, 
I'm playing with this for me and won't be submitting a pull request 
but I welcome anyone to brow,steal,take & submit what's there or a version of if they want to

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-882-a1.dts:
```diff
@@ -0,0 +1,43 @@
+	compatible = "dlink,dir-882-a1", "mediatek,mt7621-soc";
+	model = "D-Link DIR-882 A1";
+
+	leds {
```
Just use `leds: leds {` in the DTSI and the reference the label here as you do with the individual leds. Despite, there is no need for the settings the compatible again here.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-8xx-a1.dtsi:
```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
This is only required once, remove it here and keep it in the DTS files.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-8xx-a1.dtsi:
```diff
@@ -0,0 +1,174 @@
+		led_net_green: net_green {
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+		};
+	};
```
To be honest, since it's only one additional line, I'd personally prefer to keep the whole LEDs block in the DTS files. It would be much easier to read.

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -236,6 +236,21 @@ define Device/dlink_dir-860l-b1
+  IMAGE/factory.bin := append-kernel | append-rootfs | uimage-padhdr 96 |\
+	check-size
+endef
+TARGET_DEVICES += dlink_dir-867-a1
```
You can also remove the repetitions here and create a shared definition, e.g. look at what's done for Device/netgear_sercomm_nand in the same file.

adrianschmutzler commented:
As I read your initial commit, you have actually had both devices at your desk, so both are tested?

mbc07 commented:
> As I read your initial commit, you have actually had both devices at your desk, so both are tested?

Yes, both are tested. I personally own the DIR-882 and currently have access to a borrowed DIR-867 (this model is hard to come by because it's not sold anymore). I'll address your concerns and push an update to this PR very soon, thanks for the initial review.

mbc07 committed:
ramips: add support for D-Link DIR-867/DIR-882 A1

This patch adds support for D-Link DIR-867 A1 and D-Link DIR-882 A1. Given
the similarity of these devices, this patch also introduces a common DTS
shared between DIR-867 A1, DIR-878 A1 and DIR-882 A1.

Specifications:
* Board: AP-MTKH7-0002
* SoC: MediaTek MT7621AT
* RAM: 128 MB (DDR3)
* Flash: 16 MB (SPI NOR)
* WiFi: MediaTek MT7615N (x2)
* Switch: 1 WAN, 4 LAN (Gigabit)
* Ports: 1 USB 2.0, 1 USB 3.0
* Buttons: Reset, WiFi Toggle, WPS
* LEDs: Power (green/orange), Internet (green/orange), WiFi 2.4G (green),
        WiFi 5G (green), USB 2.0 (green), USB 3.0 (green)

Notes:
* WiFi 2.4G and WiFi 5G LEDs are wired directly to the wireless chips
* DIR-867 wireless chips are limited to 3x3 streams at hardware level
* USB ports and related LEDs available only on DIR-882

Serial port:
* Parameters: 57600, 8N1
* Location: J1 header (close to the Reset, WiFi and WPS buttons)
* Pinout: 1 - VCC
          2 - RXD
          3 - TXD
          4 - GND

Installation:
* D-Link Recovery GUI: power down the router, press and hold the reset
  button, then re-plug it. Keep the reset button pressed until the power
  LED starts flashing orange, manually assign a static IP address under
  the 192.168.0.xxx subnet (e.g. 192.168.0.2) and go to http://192.168.0.1

* Some modern browsers may have problems flashing via the Recovery GUI,
  if that occurs consider uploading the firmware through cURL:

    curl -v -i -F "firmware=@file.bin" 192.168.0.1

Signed-off-by: Mateus B. Cassiano <mbc07@live.com>

mbc07 commented:
Just updated the PR, let me know if there's anything else that might need attention...

adrianschmutzler commented on target/linux/ramips/dts/mt7621_dlink_dir-8xx-a1.dtsi:
```diff
@@ -0,0 +1,147 @@
+#include "mt7621.dtsi"
```
This file should have a license, too.

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -220,6 +220,19 @@ define Device/buffalo_wsr-600dhp
+define Device/dlink_dir-8xx-a1
+  IMAGE_SIZE := 16000k
+  DEVICE_VENDOR := D-Link
+  DEVICE_VARIANT := A1
```
I'd move the variant to the individual devices, as we don't know whether this will really be limited to A1, and I'd like to keep it with the DEVICE_MODEL variable.

adrianschmutzler commented:
Ah, I just realized that the removal of the SPDX in the DTSI was a misunderstanding, I only wanted you to remove the dts-v1 there. So, I added back the license and applied the DEVICE_VARIANT change myself in my staging tree.
Thanks, will be merged later today.
